README
monsti/form implements a web form generator and validator.


CHANGES
- 2014/01/04: Add DateTimeWidget, DateWidget, and TimeWidget
- 2013/01/02: Fix Form.AddError
- 2012/12/28: Add PasswordWidget
- 2012/12/07: Add FileWidget and Form.EncTypeAttr


form_test.go
// This file is part of monsti/form.
// Copyright 2012-2014 Christian Neumann

// monsti/form is free software: you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option) any
// later version.

// monsti/form is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
// details.

// You should have received a copy of the GNU Lesser General Public License
// along with monsti/form. If not, see <http://www.gnu.org/licenses/>.

package form

import (
	"html/template"
	"net/url"
	"reflect"
	"testing"
	"time"
)

type TestDataEmbed struct {
	Title string
}

type TestData struct {
	TestDataEmbed
	Name string
	Age  int
}

func TestRender(t *testing.T) {
	data := TestData{}
	form := NewForm(&data, Fields{
		"Title": Field{"Your title", "", nil, nil},
		"Name":  Field{"Your name", "Your full name", Required("Req!"), nil},
		"Age":   Field{"Your age", "Years since your birth.", Required("Req!"), nil}})
	vals := url.Values{
		"Title": []string{""},
		"Name":  []string{""},
		"Age":   []string{"14"}}
	form.Fill(vals)
	renderData := form.RenderData()
	fieldTests := []struct {
		Field    string
		Expected FieldRenderData
	}{
		{
			Field: "Title",
			Expected: FieldRenderData{
				Label:    "Your title",
				LabelTag: `<label for="Title">Your title</label>`,
				Help:     "",
				Errors:   nil,
				Input:    `<input id="Title" type="text" name="Title" value=""/>`}},
		{
			Field: "Name",
			Expected: FieldRenderData{
				Label:    "Your name",
				LabelTag: `<label for="Name">Your name</label>`,
				Help:     "Your full name",
				Errors:   []string{"Req!"},
				Input:    `<input id="Name" type="text" name="Name" value=""/>`}},
		{
			Field: "AGE",
			Expected: FieldRenderData{
				Label:    "Your age",
				LabelTag: `<label for="Age">Your age</label>`,
				Help:     "Years since your birth.",
				Errors:   nil,
				Input:    `<input id="Age" type="text" name="Age" value="14"/>`}}}
	for i, test := range fieldTests {
		if len(renderData.Errors) > 0 {
			t.Errorf("RenderData contains general errors: %v", renderData.Errors)
		}
		if !reflect.DeepEqual(renderData.Fields[i], test.Expected) {
			t.Errorf("RenderData for Field '%v' =\n%v,\nexpected\n%v",
				test.Field, renderData.Fields[i], test.Expected)
		}
	}
}

func TestAddError(t *testing.T) {
	data := TestData{}
	form := NewForm(&data, Fields{
		"Name": Field{"Your name", "Your full name", Required("Req!"), nil},
		"Age":  Field{"Your age", "Years since your birth.", Required("Req!"), nil}})
	form.AddError("Name", "Foo")
	form.AddError("", "Bar")
	renderData := form.RenderData()
	if len(renderData.Fields[0].Errors) != 1 ||
		renderData.Fields[0].Errors[0] != "Foo" {
		t.Errorf(`Field "Name" should have error "Foo"`)
	}
	if len(renderData.Errors) != 1 ||
		renderData.Errors[0] != "Bar" {
		t.Errorf(`Missing global error "Bar"`)
	}
	if len(renderData.Fields[1].Errors) != 0 {
		t.Errorf(`Field "Bar" should have no errors`)
	}
}

type TestDataEncTypeAttr struct {
	Name string
	File string
}

func TestEncTypeAttr(t *testing.T) {
	data := TestDataEncTypeAttr{}
	vals := url.Values{
		"Name": []string{""}}
	fieldTests := []struct {
		Form    *Form
		EncType string
	}{
		{
			Form: NewForm(&data, Fields{
				"Name": Field{"Your name", "Your full name", Required("Req!"),
					nil},
				"File": Field{"File Dummy", "", nil, nil}}),
			EncType: ""},
		{
			Form: NewForm(&data, Fields{
				"Name": Field{"Your name", "Your full name", Required("Req!"), nil},
				"File": Field{"File!", "", nil, new(FileWidget)}}),
			EncType: `enctype="multipart/form-data"`}}

	for i, v := range fieldTests {
		v.Form.Fill(vals)
		renderData := v.Form.RenderData()
		if string(renderData.EncTypeAttr) != v.EncType {
			t.Errorf("Test %v: RenderData.EncTypeAttr is %q, should be %q", i,
				renderData.EncTypeAttr, v.EncType)
		}
	}
}

func TestFill(t *testing.T) {
	data := TestData{}
	form := NewForm(&data, Fields{
		"Name": Field{"Your name", "Your full name", Required("Req!"), nil},
		"Age":  Field{"Your age", "Years since your birth.", Required("Req!"), nil}})
	vals := url.Values{
		"Name": []string{"Foo"},
		"Age":  []string{"14"}}
	if !form.Fill(vals) {
		t.Errorf("form.Fill(..) returns false, should be true. Errors: %v",
			form.RenderData().Errors)
	}
	vals["Name"] = []string{""}
	data.Name = ""
	if form.Fill(vals) {
		t.Errorf("form.Fill(..) returns true, should be false.")
	}
}

func TestRequire(t *testing.T) {
	invalid, valid := "", "foo"
	validator := Required("Req!")
	err := validator(valid)
	if err != nil {
		t.Errorf("require(%v) = %v, want %v", valid, err, nil)
	}
	err = validator(invalid)
	if err == nil {
		t.Errorf("require(%v) = %v, want %v", invalid, err, "'Required.'")
	}
}

func TestRegex(t *testing.T) {
	tests := []struct {
		Exp    string
		String string
		Valid  bool
	}{
		{"^[\\w]+$", "", false},
		{"^[\\w]+$", "foobar", true},
		{"", "", true},
		{"", "foobar", true},
		{"^[^!]+$", "foobar", true},
		{"^[^!]+$", "foo!bar", false}}

	for _, v := range tests {
		ret := Regex(v.Exp, "damn!")(v.String)
		if (ret == nil && !v.Valid) || (ret != nil && v.Valid) {
			t.Errorf(`Regex("%v")("%v") = %v, this is wrong!`, v.Exp, v.String,
				ret)
		}
	}
}

func TestAnd(t *testing.T) {
	tests := []struct {
		String     string
		Validators []Validator
		Valid      bool
	}{
		{"Hey! 1", []Validator{Required("Req!")}, true},
		{"", []Validator{Required("Req!")}, false},
		{"Hey! 2", []Validator{Required("Req!"), Regex("Oink", "No way!")}, false},
		{"Hey! 3", []Validator{Required("Req!"), Regex("Hey", "No way!")}, true}}
	for _, v := range tests {
		ret := And(v.Validators...)(v.String)
		if (ret == nil && !v.Valid) || (ret != nil && v.Valid) {
			t.Errorf(`And(...)("%v") = %v, this is wrong!`, v.String, ret)
		}
	}
}

func TestSelectWidget(t *testing.T) {
	widget := SelectWidget{[]Option{
		Option{"foo", "The Foo!"},
		Option{"bar", "The Bar!"}}}
	tests := []struct {
		Name, Value, Expected string
	}{
		{"TestSelect", "", `<select id="TestSelect" name="TestSelect">
<option value="foo">The Foo!</option>
<option value="bar">The Bar!</option>
</select>`},
		{"TestSelect2", "unknown!", `<select id="TestSelect2" name="TestSelect2">
<option value="foo">The Foo!</option>
<option value="bar">The Bar!</option>
</select>`},
		{"TestSelect3", "foo", `<select id="TestSelect3" name="TestSelect3">
<option value="foo" selected>The Foo!</option>
<option value="bar">The Bar!</option>
</select>`},
		{"TestSelect4", "bar", `<select id="TestSelect4" name="TestSelect4">
<option value="foo">The Foo!</option>
<option value="bar" selected>The Bar!</option>
</select>`}}
	for _, v := range tests {
		ret := widget.HTML(v.Name, v.Value)
		if string(ret) != v.Expected {
			t.Errorf(`SelectWidget.HTML("%v", "%v") = "%v", should be "%v".`,
				v.Name, v.Value, ret, v.Expected)
		}
	}
}

func TestHiddenWidget(t *testing.T) {
	widget := new(HiddenWidget)
	ret := widget.HTML("foo", "bar")
	expected := `<input id="foo" type="hidden" name="foo" value="bar"/>`
	if string(ret) != expected {
		t.Errorf(`HiddenWidget.HTML("Foo", "bar") = "%v", should be "%v".`,
			ret, expected)
	}
}

func TestFileWidget(t *testing.T) {
	widget := new(FileWidget)
	ret := widget.HTML("foo", "")
	expected := `<input id="foo" type="file" name="foo"/>`
	if string(ret) != expected {
		t.Errorf(`FileWidget.HTML("Foo", "") = "%v", should be "%v".`,
			ret, expected)
	}
}

func TestPasswordWidget(t *testing.T) {
	widget := new(PasswordWidget)
	ret := widget.HTML("foo", "")
	expected := `<input id="foo" type="password" name="foo"/>`
	if string(ret) != expected {
		t.Errorf(`PasswordWidget.HTML("Foo", "") = "%v", should be "%v".`,
			ret, expected)
	}
}

func testWidget(t *testing.T, widget Widget, data interface{}, input,
	nilInput string, value interface{}, urlValue string) {
	form := NewForm(data, Fields{"ID": Field{"T", "H", nil, widget}})
	vals := url.Values{"ID": []string{urlValue}}
	renderData := form.RenderData()
	if renderData.Fields[0].Input != template.HTML(nilInput) {
		t.Errorf("Input field for nil value is\n%v\nshould be \n%v",
			renderData.Fields[0].Input, nilInput)
	}
	form.Fill(vals)
	renderData = form.RenderData()
	if renderData.Fields[0].Input != template.HTML(input) {
		t.Errorf("Input field is\n%v\nshould be \n%v",
			renderData.Fields[0].Input, input)
	}
	if reflect.DeepEqual(reflect.ValueOf(value),
		reflect.ValueOf(data).Elem().FieldByName("ID")) {
		t.Errorf("Data is\n%v\nshould be \n%v", data, value)
	}
}

type TestDateTimeWidgetData struct {
	ID *time.Time
}

func TestDateTimeWidget(t *testing.T) {
	data := TestDateTimeWidgetData{}
	input := `<input id="ID" type="datetime" name="ID" value="2008-09-08T22:47:31-07:00"/>`
	nilInput := `<input id="ID" type="datetime" name="ID" value=""/>`
	value, err := time.Parse(time.RFC3339, "2008-09-08T22:47:31-07:00")
	if err != nil {
		t.Fatal(err)
	}
	testWidget(t, new(DateTimeWidget), &data, input, nilInput, value,
		"2008-09-08T22:47:31-07:00")
}

func TestDateWidget(t *testing.T) {
	data := TestDateTimeWidgetData{}
	input := `<input id="ID" type="date" name="ID" value="2008-09-08"/>`
	nilInput := `<input id="ID" type="date" name="ID" value=""/>`
	value, err := time.Parse("2006-01-02", "2008-09-08")
	if err != nil {
		t.Fatal(err)
	}
	testWidget(t, new(DateWidget), &data, input, nilInput, value, "2008-09-08")
}

func TestTimeWidget(t *testing.T) {
	data := TestDateTimeWidgetData{}
	input := `<input id="ID" type="time" name="ID" value="22:47:31"/>`
	nilInput := `<input id="ID" type="time" name="ID" value=""/>`
	value, err := time.Parse("15:04:05", "22:47:31")
	if err != nil {
		t.Fatal(err)
	}
	testWidget(t, new(TimeWidget), &data, input, nilInput, value, "22:47:31")
}


doc.go
// This file is part of monsti/form.
// Copyright 2012 Christian Neumann

// monsti/form is free software: you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option) any
// later version.

// monsti/form is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
// details.

// You should have received a copy of the GNU Lesser General Public License
// along with monsti/form. If not, see <http://www.gnu.org/licenses/>.

/*
Package form implements a web form generator and validator.

For each form, you have to define a struct for the form fields:
	type formData struct {
		Name string
		Age int
	}

Then you can use the form like this:
	func handle(req Request, res *Response) {
		data := formData{}
		form := form.NewForm(&data, form.Fields{
			"Name": form.Field{G("Name"), "Your Name", form.Required(G("Required.")), nil},
			"Age": form.Field{G("Age"), "Your Age",
			form.Required(G("Required.")), nil}})
		switch req.Method {
		case "GET":
			data.Name = "Default Name"
		case "POST":
			if form.Fill(req.GetFormData()) {
				save(data.Name, data.Age)
			}
		default:
			panic()
		}
		fmt.Fprint(res, renderTemplate(form.RenderData()))
	}

Fill the render data into a form template like this (html/template):
	<form action="" method="POST" accept-charset="utf-8" {{.EncTypeAttr}}>
		<fieldset>
			<div class="control-group {{if .Errors}}error{{end}}">
				<div class="controls">
					<span class="help-block"
						>{{range .Errors}}{{.}}{{end}}</span>
				</div>
			</div>
			{{range .Fields}}
			<div class="control-group {{if .Errors}}error{{end}}">
				<label class="control-label" for="name">{{.Label}}</label>
				<div class="controls">{{.Input}}
					<span class="help-block">{{.Help}}
						{{range .Errors}}{{.}}{{end}}</span></div>
			</div>
			{{end}}
			<div class="control-group">
				<div class="controls">
					<button type="submit" class="btn btn-primary">{{G "Submit"}}</button>
				</div>
			</div>
		</fieldset>
	</form>
*/
package form


form.go
// This file is part of monsti/form.
// Copyright 2012-2014 Christian Neumann

// monsti/form is free software: you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option) any
// later version.

// monsti/form is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
// details.

// You should have received a copy of the GNU Lesser General Public License
// along with monsti/form. If not, see <http://www.gnu.org/licenses/>.

package form

import (
	"fmt"
	"html"
	"html/template"
	"net/url"
	"reflect"
	"regexp"
	"strings"
	"time"

	"github.com/gorilla/schema"
)

// FieldRenderData contains the data needed for field rendering.
type FieldRenderData struct {
	// Lebel is the field's label.
	Label string
	// LabelTag is the html code for the field's label, e.g.
	// `<label for="the_id">The Label</label>`.
	LabelTag template.HTML
	// Input is the input html for the field.
	Input template.HTML
	// Help is the help string.
	Help string
	// Errors contains any validation errors.
	Errors []string
}

// RenderData contains the data needed for form rendering.
type RenderData struct {
	Fields []FieldRenderData
	Errors []string
	// EncTypeAttr is set to 'enctype="multipart/form-data"' if the Form
	// contains a File widget. Should be used as optional attribute for the form
	// element if the form may contain file input elements.
	EncTypeAttr template.HTMLAttr
}

type Widget interface {
	HTML(name string, value interface{}) template.HTML
}

// timeConverter converts a string to a time.Time
func timeConverter(in string) reflect.Value {
	out, err := time.Parse(time.RFC3339, in)
	if err != nil {
		out, err = time.Parse("2006-01-02", in)
	}
	if err != nil {
		out, _ = time.Parse("15:04:05", in)
	}
	return reflect.ValueOf(out)
}

type DateTimeWidget int

func (t DateTimeWidget) HTML(field string, value interface{}) template.HTML {
	var out string
	if obj, ok := value.(time.Time); ok {
		out = obj.Format(time.RFC3339)
	} else if obj, ok := value.(*time.Time); ok {
		if obj == nil {
			out = ""
		} else {
			out = obj.Format(time.RFC3339)
		}
	} else {
		out = fmt.Sprintf("%v", obj)
	}
	return template.HTML(fmt.Sprintf(
		`<input id="%v" type="datetime" name="%v" value="%v"/>`,
		field, field, html.EscapeString(out)))
}

type DateWidget int

func (t DateWidget) HTML(field string, value interface{}) template.HTML {
	var out string
	if obj, ok := value.(time.Time); ok {
		out = obj.Format("2006-01-02")
	} else if obj, ok := value.(*time.Time); ok {
		if obj == nil {
			out = ""
		} else {
			out = obj.Format("2006-01-02")
		}
	} else {
		out = fmt.Sprintf("%v", obj)
	}
	return template.HTML(fmt.Sprintf(
		`<input id="%v" type="date" name="%v" value="%v"/>`,
		field, field, html.EscapeString(out)))
}

type TimeWidget int

func (t TimeWidget) HTML(field string, value interface{}) template.HTML {
	var out string
	if obj, ok := value.(time.Time); ok {
		out = obj.Format("15:04:05")
	} else if obj, ok := value.(*time.Time); ok {
		if obj == nil {
			out = ""
		} else {
			out = obj.Format("15:04:05")
		}
	} else {
		out = fmt.Sprintf("%v", obj)
	}
	return template.HTML(fmt.Sprintf(
		`<input id="%v" type="time" name="%v" value="%v"/>`,
		field, field, html.EscapeString(out)))
}

type Text int

func (t Text) HTML(field string, value interface{}) template.HTML {
	return template.HTML(fmt.Sprintf(
		`<input id="%v" type="text" name="%v" value="%v"/>`,
		field, field, html.EscapeString(
			fmt.Sprintf("%v", value))))
}

type AlohaEditor int

func (t AlohaEditor) HTML(field string, value interface{}) template.HTML {
	return template.HTML(fmt.Sprintf(
		`<textarea class="editor" id="%v" name="%v"/>%v</textarea>`,
		field, field, html.EscapeString(
			fmt.Sprintf("%v", value))))
}

type TextArea int

func (t TextArea) HTML(field string, value interface{}) template.HTML {
	return template.HTML(fmt.Sprintf(
		`<textarea id="%v" name="%v"/>%v</textarea>`,
		field, field, html.EscapeString(
			fmt.Sprintf("%v", value))))
}

// Option of a select widget.
type Option struct {
	Value, Text string
}

// SelectWidget renders a selection field.
type SelectWidget struct {
	Options []Option
}

func (t SelectWidget) HTML(field string, value interface{}) template.HTML {
	var options string
	for _, v := range t.Options {
		selected := ""
		if v.Value == value.(string) {
			selected = " selected"
		}
		options += fmt.Sprintf("<option value=\"%v\"%v>%v</option>\n",
			v.Value, selected, v.Text)
	}
	ret := fmt.Sprintf("<select id=\"%v\" name=\"%v\">\n%v</select>",
		field, field, options)
	return template.HTML(ret)
}

// HiddenWidget renders a hidden input field.
type HiddenWidget int

func (t HiddenWidget) HTML(field string, value interface{}) template.HTML {
	return template.HTML(
		fmt.Sprintf(`<input id="%v" type="hidden" name="%v" value="%v"/>`,
			field, field, value))
}

// PasswordWidget renders a password field.
type PasswordWidget int

func (t PasswordWidget) HTML(field string, value interface{}) template.HTML {
	return template.HTML(
		fmt.Sprintf(`<input id="%v" type="password" name="%v"/>`,
			field, field))
}

// FileWidget renders a file upload field.
type FileWidget int

func (t FileWidget) HTML(field string, value interface{}) template.HTML {
	return template.HTML(
		fmt.Sprintf(`<input id="%v" type="file" name="%v"/>`,
			field, field))
}

// Field contains settings for a form field.
type Field struct {
	Label, Help string
	Validator   Validator
	Widget      Widget
}

// Fields is a map of field names to field settings.
type Fields map[string]Field

// Form represents an html form.
type Form struct {
	Fields map[string]Field
	data   interface{}
	errors map[string][]string
}

// NewForm creates a new Form with the given fields with data stored in the
// given pointer to a structure.
//
// In panics if data is not a pointer to a struct.
func NewForm(data interface{}, fields Fields) *Form {
	if dataType := reflect.TypeOf(data); dataType.Kind() != reflect.Ptr ||
		dataType.Elem().Kind() != reflect.Struct {
		panic("NewForm(data, fields) expects data to be a pointer to a struct.")
	}
	form := Form{data: data, Fields: fields,
		errors: make(map[string][]string, len(fields))}
	return &form
}

// RenderData returns a RenderData struct for the form.
//
// It panics if a registered field is not present in the data struct.
func (f Form) RenderData() (renderData RenderData) {
	dataVal := reflect.ValueOf(f.data).Elem()
	renderData.Fields = make([]FieldRenderData, 0, dataVal.NumField()-1)
	for name, field := range f.Fields {
		widget := field.Widget
		if widget == nil {
			widget = new(Text)
		} else if _, ok := widget.(*FileWidget); ok {
			renderData.EncTypeAttr = `enctype="multipart/form-data"`
		}
		matchFunc := func(field string) bool {
			if strings.ToLower(field) == strings.ToLower(name) {
				return true
			}
			return false
		}
		fieldVal := dataVal.FieldByNameFunc(matchFunc)
		if !fieldVal.IsValid() {
			panic("form: Registered field not present in data struct: " + name)
		}
		renderData.Fields = append(renderData.Fields, FieldRenderData{
			Label: field.Label,
			LabelTag: template.HTML(fmt.Sprintf(`<label for="%v">%v</label>`,
				name, field.Label)),
			Input:  widget.HTML(name, fieldVal.Interface()),
			Help:   field.Help,
			Errors: f.errors[name]})
	}
	renderData.Errors = f.errors[""]
	return
}

// AddError adds an error to a field's error list.
//
// To add global form errors, use an empty string as the field's name.
func (f *Form) AddError(field string, error string) {
	if f.errors[field] == nil {
		f.errors[field] = make([]string, 0, 1)
	}
	f.errors[field] = append(f.errors[field], error)
}

// Fill fills the form data with the given values and validates the form.
//
// It panics if a field has been set up which is not present in the
// data struct.
//
// Returns true iff the form validates.
func (f *Form) Fill(values url.Values) bool {
	decoder := schema.NewDecoder()
	decoder.RegisterConverter(time.Time{}, timeConverter)
	error := decoder.Decode(f.data, values)
	switch e := error.(type) {
	case nil:
		return f.validate()
	case schema.MultiError:
		for field, msg := range e {
			v, ok := f.errors[field]
			switch {
			case !ok:
				f.errors[""] = []string{msg.Error()}
			case v == nil:
				f.errors[field] = []string{msg.Error()}
			default:
				f.errors[field] = append(f.errors[field], msg.Error())
			}
		}
		return false
	default:
		panic(error.Error())
	}
}

// validate validates the currently present data.
//
// Resets any previous errors.
// Returns true iff the data validates.
func (f *Form) validate() bool {
	anyError := false
	for name, field := range f.Fields {
		value := reflect.ValueOf(f.data).Elem().FieldByName(name)
		if value == reflect.ValueOf(nil) {
			panic(fmt.Sprintf("Field '%v' not present in form data structure.",
				name))
		}
		if field.Validator != nil {
			if errors := field.Validator(value.Interface()); errors != nil {
				f.errors[name] = errors
				anyError = true
			}
		}
	}
	return !anyError
}

// Validator is a function which validates the given data and returns error
// messages if the data does not validate.
type Validator func(interface{}) []string

// And is a Validator that collects errors of all given validators.
func And(vs ...Validator) Validator {
	return func(value interface{}) []string {
		errors := []string{}
		for _, v := range vs {
			errors = append(errors, v(value)...)
		}
		if len(errors) == 0 {
			return nil
		}
		return errors
	}
}

// Required creates a Validator to check for non empty values.
//
// msg is set as validation error.
func Required(msg string) Validator {
	return func(value interface{}) []string {
		if value == reflect.Zero(reflect.TypeOf(value)).Interface() {
			return []string{msg}
		}
		return nil
	}
}

// Regex creates a Validator to check a string for a matching regexp.
//
// If the expression does not match the string to be validated,
// the given error msg is returned.
func Regex(exp, msg string) Validator {
	return func(value interface{}) []string {
		if matched, _ := regexp.MatchString(exp, value.(string)); !matched {
			return []string{msg}
		}
		return nil
	}
}
